tesseract-server/src/handlers/logic_layer/detection.rs

```diff
@@ -0,0 +1,187 @@
+    }
+    let agg_query_res = QS_NON_STRICT.deserialize_str::<AggregateQueryOpt>(&query);
+    let agg_query = match agg_query_res {
+        Ok(q) => q,
+        Err(err) => {
+            return Box::new(
+                future::result(
+                    Ok(HttpResponse::NotFound().json(err.to_string()))
+                )
+            );
+        },
+    };
+
+    // Detect cube based on the query parameters
+    let cube = detect_cube(
+        req.state().schema.read().unwrap().clone(),
+        agg_query.clone()
+    );
+    let cube = match cube {
+        Ok(cube) => cube,
+        Err(err) => {
+            return Box::new(
+                future::result(
+                    Ok(HttpResponse::NotFound().json(err.to_string()))
+                )
+            );
+        }
+    };
+    info!("cube: {:?}", cube);
+
+    finish_aggregation(req, agg_query, cube, format)
+}
+
+/// Detects which cube to use based on the drilldowns, cuts and measures provided.
+/// In case the arguments are present in more than one cube, the first cube to match all
+/// requirements is returned.
+pub fn detect_cube(schema: Schema, agg_query: AggregateQueryOpt) -> Result<String, Error> {
+    let drilldowns = match agg_query.drilldowns {
+        Some(drilldowns) => {
+            let mut d: Vec<String> = vec![];
+            for drilldown in drilldowns {
+                let e: Vec<&str> = drilldown.split(".").collect();
+                let mut final_drilldown = String::from("");
+                if e.len() == 2 {
+                    final_drilldown = format!("{}.{}.{}", e[0], e[0], e[1]).to_string();
+                } else if e.len() == 3 {
+                    final_drilldown = drilldown;
+                } else {
+                    return Err(format_err!("Wrong drilldown format. Ensure your drilldowns are correct."));
+                }
+                d.push(final_drilldown);
+            }
+            d
+        },
+        None => vec![],
+    };
+
+    let cuts = match agg_query.cuts {
+        Some(cuts) => {
+            let mut c: Vec<String> = vec![];
+            for cut in cuts {
+                let e: Vec<&str> = cut.split(".").collect();
+                let mut final_cut = String::from("");
+                if e.len() == 3 {
+                    final_cut = format!("{}.{}.{}", e[0], e[0], e[1]).to_string();
+                } else if e.len() == 4 {
+                    final_cut = format!("{}.{}.{}", e[0], e[1], e[2]).to_string();
+                } else {
+                    return Err(format_err!("Wrong cut format. Ensure your cuts are correct."));
+                }
+                c.push(final_cut);
+            }
+            c
+        },
+        None => vec![],
+    };
+
+    let measures = match agg_query.measures {
+        Some(measures) => measures,
+        None => vec![],
+    };
+
+    // TODO: Avoid clone here?
+    for cube in schema.cubes {
```

Comment:
We don't have to do this now, but the `iter().filter()` api might be nice to use.